[MPSInductor] Fix index generation for transpose

malfet opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #143966
* #143998
* #143977
* __->__ #143973
* #143949
* #143948

Alas, PythonPrinter would not work here, not would CppPrinter, so start building MetalPrinter.

`pytest test/inductor/test_torchinductor.py -k _mps` score is 474 failed, 277 passed, 32 skipped
Before this change:
`pytest test/inductor/test_torchinductor.py -k _mps` reported 506 failed, 245 passed, 32 skipped

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

malfet committed:
[MPSInductor] Fix index generation for transpose

Alas, PythonPrinter would not work here, not would CPP printer, so
invent ones own

[ghstack-poisoned]

malfet committed:
Update on "[MPSInductor] Fix index generation for transpose"


Alas, PythonPrinter would not work here, not would CPP printer, so
invent ones own

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

jansel commented on torch/_inductor/codegen/mps.py:
```diff
@@ -118,7 +144,7 @@ def store(
 
     def codegen_iteration_ranges_entry(self, entry: IterationRangesEntry) -> None:
         index_expr = self.rename_indexing(entry.expr)
-        index_str = self.sexpr(index_expr)  # type: ignore[misc]
```
Should you set `sexpr = mexpr` on the class?

malfet committed:
Update on "[MPSInductor] Fix index generation for transpose"


Alas, PythonPrinter would not work here, not would CppPrinter, so start building MetalPrinter.

`pytest test/inductor/test_torchinductor.py -k _mps` score is 474 failed, 277 passed, 32 skipped
Before this change:
`pytest test/inductor/test_torchinductor.py -k _mps` reported 506 failed, 245 passed, 32 skipped

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

malfet commented:
@pytorchbot merge

malfet commented:
@pytorchbot merge -f "Lint + Mac tests are green"
